odoo/models.py

```diff
@@ -2013,9 +2023,12 @@ def _read_group_groupby(self, groupby_spec: str, query: Query) -> SQL:
                     sql_expr, interval, interval,
                 )
             else:
-                sql_expr = SQL("date_trunc(%s, %s::timestamp)", granularity, sql_expr)
+                if granularity in RELATIVE_GROUP_TIME_GRANULARITY:
+                    sql_expr = SQL("date_part(%s, %s)", RELATIVE_GROUP_TIME_GRANULARITY.get(granularity), sql_expr)
+                else:
+                    sql_expr = SQL("date_trunc(%s, %s::timestamp)", granularity, sql_expr)
 
-            if field.type == 'date':
+            if field.type == 'date' and granularity not in RELATIVE_GROUP_TIME_GRANULARITY:
```

Comment:
Ouch, this deserves a comment: with "absolute" granularity, the expression returns a timestamp that we must convert back to a date.  With "relative" granularity, the expression returns a number instead.